src/preprocessing/passes/non_clausal_simp.cpp

```diff
@@ -448,10 +432,105 @@ PreprocessingPassResult NonClausalSimp::applyInternal(
         Rewriter::rewrite(Node(learnedBuilder)));
   }
 
-  propagator->setNeedsFinish(true);
   return PreprocessingPassResult::NO_CONFLICT;
 }  // namespace passes
 
+std::pair<bool, std::vector<Node>> NonClausalSimp::preprocessByCryptoMinisat(
+    AssertionPipeline* assertionsToPreprocess, unsigned substs_index)
+{
+  std::unique_ptr<SatSolver> satSolver(SatSolverFactory::createCryptoMinisat(
+      smtStatisticsRegistry(), "non_clausal_simp_solver"));
+  CVC4::prop::NullRegistrar registrar;
+  context::Context context;
+  CVC4::prop::TseitinCnfStream cnfStream(satSolver.get(),
+                                         &registrar,
+                                         &context,
+                                         true,
+                                         "toCNF-skeleton-preprocessing");
+  std::vector<Node> learned_literals;
+  Trace("non-clausal-simplify") << "asserting to sat solver" << std::endl;
+  {  // for timing
+    TimerStat::CodeTimer timer(d_statistics.d_cnfTranslateTime);
+    for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+    {
+      Assert(Rewriter::rewrite((*assertionsToPreprocess)[i])
+             == (*assertionsToPreprocess)[i]);
+      // Don't reprocess substitutions
+      if (substs_index > 0 && i == substs_index)
+      {
+        continue;
+      }
+      Trace("non-clausal-simplify")
+          << "asserting " << (*assertionsToPreprocess)[i] << std::endl;
+      cnfStream.convertAndAssert(
+          (*assertionsToPreprocess)[i], false, false, RULE_GIVEN);
+    }
```

Comment:
You are using `(*assertionsToPreprocess)[i]` 4 times in this loop. Maybe store it in a separate variable?